<div style="border:solid green 3px; padding: 20px">

<b>Привет, Михаил! :)</b>

Меня зовут Сороколетов Илья и я буду проверять твой проект. 

Предлагаю общаться на «**ты**», но если тебе будет комфортнее общаться на «**вы**», то сообщи об этом в комментариях)


Для твоего удобства, я буду выделять свои комментарии следующим образом:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Самые важные замечания. Они указывают на ключевые моменты, которые влияют на конечный результат проекта. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Советы или замечания, которые помогут сделать твою работу лучше, но необязательны к выполнению.

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения, например вот так:
<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
</div>

Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Если ты оставляешь свои комментарии, то делай это под моими, чтобы сообщения были расположены в хронологическом порядке. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

Обязательно задавай вопросы если они возникнут, а я перехожу к проверке)

P.S. На всякий случай, я оставлю пустой шаблон для твоих комментариев в ячейке ниже. Кликни на него два раза, скопируй код и вставляй его в пустую ячейку там, где ты хочешь оставить комментарий. Не забудь только перед этим сменить тип ячейки на Markdown. Как это сделать можно посмотреть на этом видео

    https://youtu.be/s0q2GpcYxo4?t=685

Также, перед отправкой работы я рекомендую нажимать Kernel -> Restart & Run All. Это перезапустит ядро и по очереди выполнит все ячейки. Так ты сможешь проверить, что всё работает корректно. Кнопка Kernel находится в панели сверху

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>Место для твоего комментария
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

Импорт данных в проект

In [701]:
import pandas as pd
credit_data = pd.read_csv ('/datasets/data.csv')
credit_data.to_csv('credit_data.csv', index=False)


Знакомство с данными

In [702]:
print (credit_data.head(20))
credit_data.info()
print (credit_data.columns)
print (credit_data ['children'].unique())
print (len (credit_data [credit_data['days_employed'].isna()]))
print (credit_data ['dob_years'].sort_values(ascending=True).unique())
print (credit_data ['education'].unique())
print (credit_data ['family_status'].unique())
print (credit_data ['gender'].unique())
print (credit_data ['income_type'].unique())
print (credit_data ['debt'].unique())
print (credit_data ['total_income'].min(), credit_data ['total_income'].max())
print (len (credit_data [credit_data['total_income'].isna()]))
print (credit_data ['purpose'].unique())

    children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   

**Вывод**

Количество данных в категориях не совпадает, т. е. в данных имеются пропуски. <br>Названия столбцов не имеют замечаний к стилю записи и терминологии. <br> - Категория "children" имеет целочисленный тип, но содержит отрицательные и подозрительно большие положительные значения. <br> - Категория "days_employed" имеет тип вещественных чисел и содержит положительные, отрицательные значения и пропуски. Всего 2174 пропущенных значения. Положительные значения в этой категории слишком большие для исчисления в днях поэтому можно предположить, что эти значения записаны в часах. <br> - Категория "dob_years" имеет целочисленный тип, но содержит значение 0, что противоречит здравому смыслу. Можно предположить, что 0 - значение по умолчанию при отсутствии информации в соответствующем разделе анкеты, которое автоматически заполняется самой программой.<br> - Категория "education" содержит неявные дубликаты в виде записей, выполненных в разных регистрах в следствие ручного ввода. <br> - Категория "family_status" имеет записи с замечаниями к стилю (имеются названия, начинающиеся с прописной буквы). <br> - Категория "gender" содержит нетипичные записи 'XNA'. <br> Категория "total_income", как и категория "days_employed" содержит 2174 пропуска, что может говорить о том, что данные были потеряны в результате копирования или переноса. <br> - Категория "purpose" содержит множество неявных дубликатов, записанных под разными значениями в следствие ручного ввода.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>Отличный анализ!
</div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Во время первичного анализа данных, может помочь метод describe(). Он покажет разброс значений, максимум, минимум, среднее и медиану. Ещё можно посмотреть долю пропущенных значений среди всех данных
    
    df.isna().mean()
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

Столбец 'total_income' имеет пропуски. Ранее мы выяснили, что таких строк 2174 из 21525, что составляет более 10%. Заменим эти значения медианным значением, которое расчитаем, разбив всех потребетиелей на возрастные группы по столбцу 'dob_year'. Однако данный столбец содержит значение 0, которое противоречит здравому смыслу и, скорее всего, было подставлено программой автоматически в результате пропуска информации в данном поле при ручном вводе. Выясним сколько записей содержат 0 в графе 'dob_years'

In [703]:
print (len(credit_data[credit_data['dob_years'] == 0]))

101


Всего 101 запись. Это составляет около 0,47% от общего числа записей. Таким образом эти данные могут быть удалены, чтобы не оказывать влияние выводы.

In [704]:
credit_data = credit_data[credit_data['dob_years'] != 0]
print (len(credit_data[credit_data['dob_years'] == 0]))

0


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>👍
</div>


Разобъём заёмщков на группы по возрастам и заменим пропуски в столбце 'total_income'на медианные значения для каждой группы.

In [705]:
print(len(credit_data[credit_data['total_income'].isna()]))

index = credit_data.index
age_groups = [[25, 30], [30, 35], [35, 40], [40, 45], [45, 50], [50, 55], [55, 60], [60, 65]]
for group in age_groups:
    credit_data.loc[(credit_data['dob_years'] > group[0]) & (credit_data['dob_years'] <= group[1]) \
                    & (credit_data['days_employed'].isna() == True),'total_income'] \
    = credit_data[(credit_data['dob_years'] > group[0]) \
                  & (credit_data['dob_years'] <= group[1])]['total_income'].median()
    
    condition = credit_data[(credit_data['dob_years'] > group[0]) & (credit_data['dob_years'] <= group[1])].index
    credit_data.loc[condition,'mean_age'] = (group[0] + group[1]) / 2

    
credit_data.loc[(credit_data['dob_years'] <= 25) & (credit_data['days_employed'].isna() == True),'total_income'] \
    = credit_data[(credit_data['dob_years'] <= 25)]['total_income'].median()
credit_data.loc[credit_data['dob_years'] <= 25,'mean_age'] = credit_data[credit_data['dob_years'] <= 25]['dob_years'].mean()
 
credit_data.loc[(credit_data['dob_years'] > 65) & (credit_data['days_employed'].isna() == True),'total_income'] \
    = credit_data[(credit_data['dob_years'] > 65)]['total_income'].median()
credit_data.loc[credit_data['dob_years'] > 65,'mean_age'] = credit_data[credit_data['dob_years'] > 65]['dob_years'].mean()

credit_data['mean_age'] = credit_data['mean_age'].astype('int')

#print (credit_data['mean_age'].sort_values().unique())

print(len(credit_data[credit_data['total_income'].isna()]))
#print (credit_data)

2164
0


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Здесь лучше заменить пропуски на медиану по типу занятости, так как она лучше позволяет более точно восстановить пропущенные значения
</div>

Аналогичным образом поступим с пропущенными данными в столбце 'days_employed'. Заменим их медианными значениями, разбив заёмщиков на группы по типу дохода 'income_type'. Но т.к. в этой категории встречаются и отрицательные, и положительные значения, последние из которых, по всей видимости, выражены не в днях, а в часах, то сначала переведём все значения их часов в дни, а затем сделаем все значения положительными.

In [706]:
print(len(credit_data.loc[credit_data['days_employed'] > 0, 'days_employed']))
print(credit_data.loc[credit_data['days_employed'] > 0, 'days_employed'].max())
credit_data.loc[credit_data['days_employed'] > 0, 'days_employed'] = \
credit_data.loc[credit_data['days_employed'] > 0, 'days_employed'] / 24
print(len(credit_data.loc[credit_data['days_employed'] > 0, 'days_employed']))
print(credit_data.loc[credit_data['days_employed'] > 0, 'days_employed'].max())

3428
401755.40047533
3428
16739.80835313875


In [707]:
print(len(credit_data[credit_data['days_employed'] < 0]))
credit_data.loc[credit_data['days_employed'] < 0, 'days_employed'] = \
credit_data.loc[credit_data['days_employed'] < 0, 'days_employed'] * -1
print(len(credit_data[credit_data['days_employed'] < 0]))

15832
0


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Тут можно использовать функцию abs() для отрицательных значений, которая возвращает модуль от числа
</div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Делить все значения на 24 часа не совсем правильно. Так нужно поступить только с аномально высоким стажем, который выше возраста - 18.    
</div>

In [708]:
income_type = ['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете']
for index in income_type:
    credit_data.loc[(credit_data['income_type'] == index) & (credit_data['days_employed'].isna() == True),'days_employed'] \
    = credit_data[credit_data['income_type'] == index]['days_employed'].median()
#    print (credit_data[credit_data['income_type'] == index]['days_employed'].median())

credit_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21424 entries, 0 to 21524
Data columns (total 13 columns):
children            21424 non-null int64
days_employed       21424 non-null float64
dob_years           21424 non-null int64
education           21424 non-null object
education_id        21424 non-null int64
family_status       21424 non-null object
family_status_id    21424 non-null int64
gender              21424 non-null object
income_type         21424 non-null object
debt                21424 non-null int64
total_income        21424 non-null float64
purpose             21424 non-null object
mean_age            21424 non-null int64
dtypes: float64(2), int64(6), object(5)
memory usage: 2.9+ MB


После заполнения пропущенных значений повторно проверим данные на наличие пропусков

In [709]:
print (credit_data.isna().sum())

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
mean_age            0
dtype: int64


Для выполнения поставленных задач: "Есть ли зависимость между наличием детей и возвратом кредита в срок?" и "Есть ли зависимость между семейным положением и возвратом кредита в срок?" так же важны данные из категорий 'children' и 'purpose'. В категории 'children' есть нетипично большие значения 20.<br> Выясним количество таких значений и их расположение в таблице

In [710]:
print (credit_data[credit_data['children'] == 20])
print (credit_data[credit_data['children'] == 20]['children'].count())

       children  days_employed  dob_years education  education_id  \
606          20     880.221113         21   среднее             1   
720          20     855.595512         44   среднее             1   
1074         20    3310.411598         56   среднее             1   
2510         20    2714.161249         59    высшее             0   
3302         20    2673.404956         35   среднее             1   
...         ...            ...        ...       ...           ...   
21008        20    1240.257910         40   среднее             1   
21325        20     601.174883         37   среднее             1   
21390        20    1548.637544         53   среднее             1   
21404        20     494.788448         52   среднее             1   
21491        20     173.954460         27   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
606          женат / замужем                 0      M    компаньон     0   
720          женат 

Всего нашлось 75 из 21424 строк с нетипично большим значением, которые расположены в разных частях таблицы и имеют одинаковый с остальными значениями тип данных. Вероятнее всего эти значения - промежуточный итог в группировках по какому-то критерию. Таким образом нельзя однозначно определить истинные значения для данных строк и поскольку доля таких значений составляет мнее 0,35%, то эти значения удалим для исключения возможного влияния на результаты.

In [711]:
credit_data = credit_data[credit_data['children'] != 20]

print (credit_data[credit_data['children'] == 20]['children'].count())

0


<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Как вариант, 20 можно было бы заменить на 2, так как в итоговых вопросах нас интересует именно наличие детей, а не их количество
</div>

Категория 'gender' содержит значения 'XNA'. Подсчитаем количество этих значений и выведем строки с таким значением для анализа.

In [712]:
print (credit_data[credit_data['gender'] == 'XNA']['gender']. count())
print (credit_data.loc[credit_data['gender'] == 'XNA'])

1
       children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  mean_age  
10701  203905.157261  покупка недвижимости        23  


В данном случае пол не указан всего в одном варианте и это, вероятнее всего, пропуск данных в анкете, который был заменён программой на значение по умолчанию 'XNA' для отсутствующих данных. Отсутствие указания пола в данных не повлияет на исследования, поэтому оставим эту запись без изменений.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b>
<br>Поддерживаю! Так как мы не можем достоверно сказать, к какому полу ближе это значение, его стоит или удалить или не трогать)
</div>

**Вывод**

В таблице обнаружились пропуски и нетипичные данные, которые могли повлиять на результаты исследования. Пропуски в категориях 'days_employed' и 'total_income' совпадали по количеству, что может говорить о потере данных в результате копирования. Поскольку общее количество пропусков составило более 10%, пропущенные значения были заменены на медианные значения для определённых возрастных групп заёмщиков в результате чего таблица стала пригодна для дальнейшего анализа.

### Замена типа данных

Категория 'days_employed' имеет тип вещественных чисел. Для исследований будет удобнее перевести эти данные в целочисленный тип.

In [713]:
credit_data['days_employed'] = credit_data['days_employed'].astype('int')
credit_data['days_employed'].dtype

dtype('int64')

Категория 'total_income' имеет вещественный тип данных с точностью до 6 знаков после запятой, которая не требуется для текущего анализа. <br>
Переведём данные в этой категории в целочисленный тип для экономии памяти и времени вычислительных операций.

In [714]:
credit_data['total_income'] = credit_data['total_income'].astype('int')
credit_data['total_income'].dtype

dtype('int64')

**Вывод**

Тип данных в колонках 'days_employed' и 'total_income' изменён для удобства дальнейшей работы с данными.

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>А как же столбец total_income? В нём значения с шестью знаками после запятой)
</div>

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>

<br> total_income перевёл в целочисленный тип


### Обработка дубликатов

Рассмотрим  по порядку значения в столбцах на предмет явных и неявных дубликатов. <br>- В столбце 'children' есть отрицательные значения. Они, скорее всего, являются артефактами ручного ввода с использованием "-" перед числом, поэтому заменяем отрицательные числа на положительные, эквивалентные по модулю.<br>- В категории 'day_employed' ранее мы заполнили пропуски (на всякий случай убедимся ещё раз в этом) и перевели значения, указанные в часах, в дни, а так же заменили все отрицательные значения на положительные. <br>- В категории 'dob_years' ранее мы исключили записи со значением 0 (на всякий случай убедимся ещё раз).<br>- Категория "education" содержит неявные дубликаты в виде записей, выполненных в разных регистрах в следствие ручного ввода - переведём все записи в ней в нижний регистр.

In [715]:
credit_data ['children'] = credit_data['children'].replace (-1,1)

print (credit_data ['children'].sort_values().unique())
print (credit_data['days_employed'].isna().sum())
print (credit_data['dob_years'].sort_values().unique())

credit_data.loc[:,'education'] = credit_data.loc[:,'education'].str.lower()
print (credit_data['education'].unique())


[0 1 2 3 4 5]
0
[19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42
 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66
 67 68 69 70 71 72 73 74 75]
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']


**Вывод**

Данные подготовлены для дальнейших лемматизации и категоризации.

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Стоит проверить данные на наличие дубликатов связкой duplicated().sum() и после этого удалить их
</div>

<div class="alert alert-info">
<b> Комментарий студента:</b> 
<br> Не могу выполнить данный запрос, т.к. в перечне данных отсутсвует категория с уникальными данными (id или ФИО заёмщиков), подлежащими проверке duplicated().sum() на наличие явных дубликатов (в остальных категориях данные могут повторяться, т.к. отражают состояние из предопределённых статусов: образование, тип занятости, семейное положение и т.п. или состоят из непрерывных случайных величин: доход, продолжительность работы, возраст, которые в силу случайной природы происхождения могут повторяться)<br>
Все категории, содержащие качественные значения были ранее проверены на наличие неявных дубликатов методом unique. Категория 'purpose' содержит множество неявных дубликатов, которые будут удалены с помощью лемматизации в следующем разделе. 
</div>

### Лемматизация

Раздел данных 'purpose' содержит множество неявных дубликатов в следствие ручного ввода. Ознакомимся со списком уникальных значений и применим лемматизацию к данной категории.

In [716]:
print (credit_data['purpose'].unique())

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


In [717]:
from pymystem3 import Mystem
m = Mystem()
from collections import Counter
 
lem_list = [] 
for word in credit_data['purpose']:
    lemmas = ' '.join(m.lemmatize(word))
    lem_list.append(lemmas)
#    print(lemmas)
#    print(Counter(lemmas))

credit_data['lemmas'] = lem_list
#print (lem_list)
#print (credit_data)
print (pd.Series(lem_list).unique())

['покупка   жилье \n' 'приобретение   автомобиль \n'
 'дополнительный   образование \n' 'сыграть   свадьба \n'
 'операция   с   жилье \n' 'образование \n' 'на   проведение   свадьба \n'
 'покупка   жилье   для   семья \n' 'покупка   недвижимость \n'
 'покупка   коммерческий   недвижимость \n'
 'покупка   жилой   недвижимость \n'
 'строительство   собственный   недвижимость \n' 'недвижимость \n'
 'строительство   недвижимость \n'
 'на   покупка   подержать   автомобиль \n'
 'на   покупка   свой   автомобиль \n'
 'операция   с   коммерческий   недвижимость \n'
 'строительство   жилой   недвижимость \n' 'жилье \n'
 'операция   со   свой   недвижимость \n' 'автомобиль \n'
 'заниматься   образование \n' 'сделка   с   подержанный   автомобиль \n'
 'получение   образование \n' 'свадьба \n'
 'получение   дополнительный   образование \n' 'покупка   свой   жилье \n'
 'операция   с   недвижимость \n' 'получение   высокий   образование \n'
 'свой   автомобиль \n' 'сделка   с   автомобиль \n'
 'про

**Вывод**

Лемматизация произведена, значения лемм добавлены в дополнительный столбик 'lemmas' основной таблицы. Данные готовы к следующему этапу категоризации.

### Категоризация данных

Выделим из полученных лемм 5 основных категорий кредитов и добавим соответствующий столбик в общую таблицу

In [718]:
lem_groups = ['ремонт', 'автомобиль', 'образование', 'свадьба', 'коммерческий', 'жилье']     


index = credit_data.index
condition = credit_data['lemmas'].str.contains(lem_groups[0])
applies_indexes = index[condition]
credit_data.loc[applies_indexes,'purpose_type'] = 'ремонт'
condition = credit_data['lemmas'].str.contains(lem_groups[1])
applies_indexes = index[condition]
credit_data.loc[applies_indexes,'purpose_type'] = 'автомобиль'
condition = credit_data['lemmas'].str.contains(lem_groups[2])
applies_indexes = index[condition]
credit_data.loc[applies_indexes,'purpose_type'] = 'образование'
condition = credit_data['lemmas'].str.contains(lem_groups[3])
applies_indexes = index[condition]
credit_data.loc[applies_indexes,'purpose_type'] = 'свадьба'
condition = credit_data['lemmas'].str.contains(lem_groups[4])
applies_indexes = index[condition]
credit_data.loc[applies_indexes,'purpose_type'] = 'коммерческая недвижимость'
credit_data['purpose_type'] = credit_data['purpose_type'].fillna('жильё')


#print (credit_data['purpose_type'].isna().sum())
print (credit_data.head(5))


#credit_data.info()

   children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  mean_age                           lemmas  \
0               покупка жилья        42             

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> 
<br>Если ты посмотришь повнимательнее, то заметишь, что коммерческий связано с недвижимостью, соотвественно нужно или поменять формулировку на "коммерческая недвижимость" или объеденить с категорией "жильё"
</div>

Разобьём заёмщиков на три категории по доходам: 'низкий', 'средний', 'высокий'. <br>
Для этого определим среднее значение доходов, медиану, минимум и максимум. Затем численно подберём пороговые значения среднего дохода так, чтобы все данные разбились на примерно равные по количеству заёмщиков категории.

In [719]:
print (credit_data['total_income'].min())
print (credit_data['total_income'].median())
print (credit_data['total_income'].mean())
print (credit_data['total_income'].max())
print (credit_data['total_income'].count())
print (credit_data[(credit_data['total_income'] > 122500) \
                   & (credit_data['total_income'] <= credit_data['total_income'].median())]['total_income'].count())
print (credit_data[(credit_data['total_income'] > credit_data['total_income'].median()) \
                   & (credit_data['total_income'] <= 170500 )]['total_income'].count())


20667
147374.0
165268.55281277813
2265604
21349
3554
3558


Добавим соответствующие категории в общую таблицу.

In [720]:
credit_data.loc[credit_data['total_income'] < 122500,'income_rank'] = 'низкий'
credit_data.loc[credit_data['total_income'] > 170500,'income_rank'] = 'высокий'
credit_data['income_rank'] = credit_data['income_rank'].fillna('средний')
print (credit_data['income_rank'].isna().sum())

0


<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> 
<br>Ты правильно и логично выделил категории по уровню дохода. В реальной задаче лучше уточнить у заказчика или коллег как лучше это будет сделать
</div>

<br>

**Вывод**

Данные разбиты на категории для дальнейшего анализа

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [721]:
cd_children_pivot = credit_data.pivot_table(index=['children'], columns='debt', values='total_income', aggfunc='count')

for index in cd_children_pivot:
    
    try:
        cd_children_pivot['ratio'] = cd_children_pivot [0] / cd_children_pivot [1]
        
    except:
        cd_children_pivot['ratio'] = NaN

final_children_presence_ratio = cd_children_pivot.loc[1:,0].sum() / cd_children_pivot.loc[1:,1].sum()

print (cd_children_pivot)
print ('Соотношение возвращённых в срок кредитов к невозвращённым в срок для заёмщиков без детей', cd_children_pivot.loc[0,'ratio'])
print ('Соотношение возвращённых в срок кредитов к невозвращённым в срок для заёмщиков с детьми', final_children_presence_ratio)

debt            0       1      ratio
children                            
0         13022.0  1058.0  12.308129
1          4407.0   442.0   9.970588
2          1848.0   194.0   9.525773
3           301.0    27.0  11.148148
4            37.0     4.0   9.250000
5             9.0     NaN        NaN
Соотношение возвращённых в срок кредитов к невозвращённым в срок для заёмщиков без детей 12.30812854442344
Соотношение возвращённых в срок кредитов к невозвращённым в срок для заёмщиков с детьми 9.898050974512744


<div class="alert alert-info">
<b> Комментарий студента:</b> 
<br>
    <br>
    Новый расчёт ниже:

In [722]:
cd_children_pivot_narrow  = credit_data.groupby(['children','debt']).agg({'total_income':'count'})
print (cd_children_pivot_narrow)

cd_children_pivot_narrow = cd_children_pivot_narrow.reset_index(level = 'children') #реиндексация сводной таблицы для фильтрации по наличию долга и детей

#количество клиентов без детей и с долгом / (количество клиентов без детей и с долгом + количество клиентов без детей и без долга)
no_children_debt_to_total_no_children_ratio = cd_children_pivot_narrow[(cd_children_pivot_narrow.index == 1) & \
                             (cd_children_pivot_narrow['children'] == 0)]['total_income'].sum() / \
                                cd_children_pivot_narrow[cd_children_pivot_narrow['children'] == 0]\
                                    ['total_income'].sum()

#количество клиентов с детьми с долгами / (количество клиентов с детьми и с долгами + количество клиентов с детьми без долгов)
with_children_debt_to_total_with_children_ratio = \
    cd_children_pivot_narrow[(cd_children_pivot_narrow.index == 1) & \
                             (cd_children_pivot_narrow['children'] > 0)]['total_income'].sum() / \
                                cd_children_pivot_narrow[cd_children_pivot_narrow['children'] > 0]\
                                    ['total_income'].sum() 

print ('Соотношение кол-ва клиентов без детей с долгами к общему кол-ву клиентов без детей {:.2%}'.format (no_children_debt_to_total_no_children_ratio))
print ('Соотношение кол-ва клиентов c детьми с долгами к общему кол-ву клиентов с детьми {:.2%}'.format (with_children_debt_to_total_with_children_ratio))

#print(cd_children_pivot_narrow)



               total_income
children debt              
0        0            13022
         1             1058
1        0             4407
         1              442
2        0             1848
         1              194
3        0              301
         1               27
4        0               37
         1                4
5        0                9
Соотношение кол-ва клиентов без детей с долгами к общему кол-ву клиентов без детей 7.51%
Соотношение кол-ва клиентов c детьми с долгами к общему кол-ву клиентов с детьми 9.18%


**Вывод**

Согласно полученным данным заёмщики без детей чаще возвращают кредиты в срок.

<div class="alert alert-info">
<b> Комментарий студента:</b> 

UPD: Среди клиентов без детей задержки по кредиту имеют 7.51% заёмщиков, тогда как среди заёмщиков с детьми задержки по кредиту имеют 9.18% <br>
Что, в свою очередь, подтверждает сделанные ранее выводы.
    
 </div>

<br>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [723]:
cd_family_status_pivot = credit_data.pivot_table(index=['family_status'], columns='debt', values='total_income', aggfunc='count')

for index in cd_family_status_pivot:
    
    try:
        cd_family_status_pivot['ratio'] = cd_family_status_pivot [0] / cd_family_status_pivot [1]
        
    except:
        cd_family_status_pivot['ratio'] = NaN

print (cd_family_status_pivot.sort_values('ratio'))

debt                       0    1      ratio
family_status                               
Не женат / не замужем   2516  272   9.250000
гражданский брак        3761  383   9.819843
женат / замужем        11359  924  12.293290
в разводе               1099   84  13.083333
вдовец / вдова           889   62  14.338710


<div class="alert alert-info">
<b> Комментарий студента:</b> 
<br>
    <br>
    Новый расчёт ниже:

In [724]:
cd_family_status_pivot = cd_family_status_pivot.reset_index(level = 'family_status') #реиндексация сводной таблицы

In [725]:
#количество одиноких клиентов с долгами / (количество одиноких клиентов с долгами + количество одиноких клиентов без долгов)
single_clients_debt_to_total_single_clients_ratio = \
 cd_family_status_pivot.loc[cd_family_status_pivot['family_status'] == 'Не женат / не замужем', 1].sum() / \
 (cd_family_status_pivot.loc[cd_family_status_pivot['family_status'] == 'Не женат / не замужем', 0].sum() + \
  cd_family_status_pivot.loc[cd_family_status_pivot['family_status'] == 'Не женат / не замужем', 1].sum())

#количество не одиноких клиентов с долгами / (количество не одиноких клиентов с долгами + количество не одиноких клиентов без долгов)
not_single_clients_debt_to_total_not_single_clients_ratio = \
cd_family_status_pivot.loc[cd_family_status_pivot['family_status'] != 'Не женат / не замужем', 1].sum() / \
(cd_family_status_pivot.loc[cd_family_status_pivot['family_status'] != 'Не женат / не замужем', 0].sum() + \
  cd_family_status_pivot.loc[cd_family_status_pivot['family_status'] != 'Не женат / не замужем', 1].sum())


#print (cd_family_status_pivot)
print ('Соотношение кол-ва одиноких клиентов с долгами к общему кол-ву одиноких клиентов {:.2%}'.format (single_clients_debt_to_total_single_clients_ratio))
print ('Соотношение кол-ва не одиноких клиентов с долгами к общему кол-ву не одиноких клиентов {:.2%}'.format (not_single_clients_debt_to_total_not_single_clients_ratio))
                           

Соотношение кол-ва одиноких клиентов с долгами к общему кол-ву одиноких клиентов 9.76%
Соотношение кол-ва не одиноких клиентов с долгами к общему кол-ву не одиноких клиентов 7.83%


**Вывод**

Согласно полученным данным заёмщики не состоящие в браке или состоящие в гражданском браке реже возвращают кредиты в срок, чем заёмщики состоящие или состоявщие в отношениях.

<div class="alert alert-info">
<b> Комментарий студента:</b> 

UPD: Среди одиноких клиентов задержки по кредиту имеют 9.76% заёмщиков, тогда как среди неодиноких заёмщиков задержки по кредиту имеют 7.83% <br>
Что, в свою очередь, подтверждает сделанные ранее выводы.
    
 </div>

<br>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [726]:
cd_income_rank_pivot = credit_data.pivot_table(index=['income_rank'], columns='debt', values='total_income', aggfunc='count')

for index in cd_income_rank_pivot:
    
    try:
        cd_income_rank_pivot['ratio'] = cd_income_rank_pivot [0] / cd_income_rank_pivot [1]
        
    except:
        cd_income_rank_pivot['ratio'] = NaN

print (cd_income_rank_pivot.sort_values('ratio'))


#income_rank_grouped = credit_data.groupby(['income_rank','debt']).agg({'total_income':'count'})

#print(income_rank_grouped)

debt            0    1      ratio
income_rank                      
средний      6506  606  10.735974
низкий       6541  581  11.258176
высокий      6577  538  12.224907


<div class="alert alert-info">
<b> Комментарий студента:</b> 
<br>
    <br>
    Новый расчёт ниже:

In [727]:
cd_income_rank_pivot = cd_income_rank_pivot.reset_index(level = 'income_rank') #реиндексация сводной таблицы

In [728]:
#количество клиентов с низким доходом с долгами / (количество клиентов с низким доходом с долгами + количество клиентов с низким дохдом без долгов)
low_income_rank_debt_clients_ratio = \
 cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'низкий', 1].sum() / \
 (cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'низкий', 0].sum() + \
  cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'низкий', 1].sum())

#количество клиентов со средним доходом с долгами / (количество клиентов со средним доходом с долгами + количество клиентов со средним дохдом без долгов)
medium_income_rank_debt_clients_ratio = \
 cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'средний', 1].sum() / \
 (cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'средний', 0].sum() + \
  cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'средний', 1].sum())

#количество клиентов с высоким доходом с долгами / (количество клиентов с высоким доходом с долгами + количество клиентов с высоким дохдом без долгов)
high_income_rank_debt_clients_ratio = \
 cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'высокий', 1].sum() / \
 (cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'высокий', 0].sum() + \
  cd_income_rank_pivot.loc[cd_income_rank_pivot['income_rank'] == 'высокий', 1].sum())



#print (cd_income_rank_pivot)
print ('Соотношение кол-ва клиентов с низким уровнем дохода с долгами к общему кол-ву клиентов с низким уровнем дохода {:.2%}'.format (low_income_rank_debt_clients_ratio))
print ('Соотношение кол-ва клиентов со средним уровнем дохода с долгами к общему кол-ву клиентов со средним уровнем дохода {:.2%}'.format (medium_income_rank_debt_clients_ratio))
print ('Соотношение кол-ва клиентов с высоким уровнем дохода с долгами к общему кол-ву клиентов с высоким уровнем дохода {:.2%}'.format (high_income_rank_debt_clients_ratio))                           

Соотношение кол-ва клиентов с низким уровнем дохода с долгами к общему кол-ву клиентов с низким уровнем дохода 8.16%
Соотношение кол-ва клиентов со средним уровнем дохода с долгами к общему кол-ву клиентов со средним уровнем дохода 8.52%
Соотношение кол-ва клиентов с высоким уровнем дохода с долгами к общему кол-ву клиентов с высоким уровнем дохода 7.56%


**Вывод**

Заёмщики с высоким доходом чаще других возвращают кредиты вовремя. Заёмщики со средним доходом реже других возвращают кредиты вовремя, однако разница по отношению к другим категориям заёмщиком невелика.<br> <br> 


<div class="alert alert-info">
<b> Комментарий студента:</b> 

UPD: Среди клиентов со средним доходом задержки по кредиту имеют 8.52% заёмщиков, тогда как среди заёмщиков с низким доходом задержки по кредиту имеют 8.16%, а среди заёмщиков с высоким доходом задержки по кредиту имеют 7.56%. Таким образом чаще всех возвращают кредиты в срок клиенты с высоким уровнем дохода, а реже всех клиенты со средним уровнем дохода.<br>
Что, в свою очередь, подтверждает сделанные ранее выводы.
    
 </div>

- Как разные цели кредита влияют на его возврат в срок?

In [729]:
cd_purpose_pivot = credit_data.pivot_table(index=['purpose_type','purpose'], columns='debt', values='total_income', aggfunc='count')
cd_purpose_pivot_reset_index = cd_purpose_pivot.reset_index ()

purpose_grouped = credit_data.groupby(['purpose_type','debt']).agg({'total_income':'count'})

#print(cd_purpose_pivot)
print(purpose_grouped)

                                total_income
purpose_type              debt              
автомобиль                0             3880
                          1              398
жильё                     0             8202
                          1              644
коммерческая недвижимость 0             1205
                          1               98
образование               0             3620
                          1              369
ремонт                    0              573
                          1               35
свадьба                   0             2144
                          1              181


<div class="alert alert-info">
<b> Комментарий студента:</b> 
<br>
    <br>
    Новый расчёт ниже:

In [730]:
purpose_grouped = purpose_grouped.reset_index(level = 'debt') #реиндексация сводной таблицы для фильтрации по наличию долга и детей

In [731]:
#количество клиентов с долгом, взявших кредит на автомобиль / (количество клиентов с долгом, взявших кредит на автомобиль + количество клиентов без долга, взявших кредит на автомобиль)
vehicle_credit_ratio = purpose_grouped[(purpose_grouped.index == 'автомобиль') & \
     (purpose_grouped['debt'] == 1)]['total_income'].sum() / \
       purpose_grouped[purpose_grouped.index == 'автомобиль']['total_income'].sum()

#количество клиентов с долгом, взявших кредит на жильё / (количество клиентов с долгом, взявших кредит на жильё + количество клиентов без долга, взявших кредит на жильё)
accomodation_credit_ratio = purpose_grouped[(purpose_grouped.index == 'жильё') & \
     (purpose_grouped['debt'] == 1)]['total_income'].sum() / \
       purpose_grouped[purpose_grouped.index == 'жильё']['total_income'].sum()

#количество клиентов с долгом, взявших кредит коммерческую недвижимость / (количество клиентов с долгом, взявших кредит на коммерческую недвижимость + количество клиентов без долга, взявших кредит на коммерческую недвижимость)
real_estate_credit_ratio = purpose_grouped[(purpose_grouped.index == 'коммерческая недвижимость') & \
     (purpose_grouped['debt'] == 1)]['total_income'].sum() / \
       purpose_grouped[purpose_grouped.index == 'коммерческая недвижимость']['total_income'].sum()

#количество клиентов с долгом, взявших кредит на образование / (количество клиентов с долгом, взявших кредит на образование + количество клиентов без долга, взявших кредит на образование)
education_credit_ratio = purpose_grouped[(purpose_grouped.index == 'образование') & \
     (purpose_grouped['debt'] == 1)]['total_income'].sum() / \
       purpose_grouped[purpose_grouped.index == 'образование']['total_income'].sum()

#количество клиентов с долгом, взявших кредит на ремонт / (количество клиентов с долгом, взявших кредит на ремонт + количество клиентов без долга, взявших кредит на ремонт)
rework_credit_ratio = purpose_grouped[(purpose_grouped.index == 'ремонт') & \
     (purpose_grouped['debt'] == 1)]['total_income'].sum() / \
       purpose_grouped[purpose_grouped.index == 'ремонт']['total_income'].sum()

#количество клиентов с долгом, взявших кредит на свадьбу / (количество клиентов с долгом, взявших кредит на свадьбу + количество клиентов без долга, взявших кредит на свадьбу)
wedding_credit_ratio = purpose_grouped[(purpose_grouped.index == 'свадьба') & \
     (purpose_grouped['debt'] == 1)]['total_income'].sum() / \
       purpose_grouped[purpose_grouped.index == 'свадьба']['total_income'].sum()



print ('Соотношение кол-ва клиентов взявших кредит на автомобиль с долгами к общему кол-ву клиентов взявших кредит на автомобиль {:.2%}'.format (vehicle_credit_ratio))
print ('Соотношение кол-ва клиентов взявших кредит на жильё с долгами к общему кол-ву клиентов взявших кредит на жильё {:.2%}'.format (accomodation_credit_ratio))
print ('Соотношение кол-ва клиентов взявших кредит на коммерческую недвижимость с долгами к общему кол-ву клиентов взявших кредит на коммерческую недвижимость {:.2%}'.format (real_estate_credit_ratio))
print ('Соотношение кол-ва клиентов взявших кредит на обучение с долгами к общему кол-ву клиентов взявших кредит на обучение {:.2%}'.format (education_credit_ratio))
print ('Соотношение кол-ва клиентов взявших кредит на ремонт с долгами к общему кол-ву клиентов взявших кредит на ремонт {:.2%}'.format (rework_credit_ratio))
print ('Соотношение кол-ва клиентов взявших кредит на свадьбу с долгами к общему кол-ву клиентов взявших кредит на свадьбу {:.2%}'.format (wedding_credit_ratio))

#print(purpose_grouped)





Соотношение кол-ва клиентов взявших кредит на автомобиль с долгами к общему кол-ву клиентов взявших кредит на автомобиль 9.30%
Соотношение кол-ва клиентов взявших кредит на жильё с долгами к общему кол-ву клиентов взявших кредит на жильё 7.28%
Соотношение кол-ва клиентов взявших кредит на коммерческую недвижимость с долгами к общему кол-ву клиентов взявших кредит на коммерческую недвижимость 7.52%
Соотношение кол-ва клиентов взявших кредит на обучение с долгами к общему кол-ву клиентов взявших кредит на обучение 9.25%
Соотношение кол-ва клиентов взявших кредит на ремонт с долгами к общему кол-ву клиентов взявших кредит на ремонт 5.76%
Соотношение кол-ва клиентов взявших кредит на свадьбу с долгами к общему кол-ву клиентов взявших кредит на свадьбу 7.78%


**Вывод**

Больше всего кредитов берут на операции с жильём. В этой же категории кредиты чаще других возвращают в срок. <br>
Чаще всего заёмщики допускают просрочки по кредитам, которые берут для сделок с автомобилями.

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> 
<br>Во всех четырёх вопросах тебе нужно поправить расчеты:
    
    количество клиентов без детей и с долгом / (количество клиентов без детей и с долгом + количество клиентов без детей и без долга)
    
<br>Такие расчеты нужно произвести для каждой категории клиентов во всех четырёх вопросах. Лучше всего для этого построить сводные таблицы используя метод pivot_table
</div>

<div class="alert alert-info">
<b> Комментарий студента:</b> 

UPD: Среди клиентов, взявших потребительские кредиты на различные цели чаще других не возвращают во время 9.3% клиентов в своей категории, взявших кредит на покупку автомобиля. В остальных категриях не возвращают в срок: <br>
    - 9.25% клиентов, взявших кредит на образование <br>
    - 7.78% клиентов, взявших кредит на свадьбу <br>
    - 7.52% клиентов, взявших кредит на покупку коммерческой недвижимости <br>
    - 7.28% клиентов, взявших кредит на покупку жилья <br>
    - 5.76% клиентов, взявших кредит на ремонт. <br>
<br>
В приведённых ранее выводах указывалось, что реже всего в срок заёмщики возвращают в срок кредиты взятые на покупку автомобиля и новые данные подтвердили этот факт. Однако разница в просрочках со следующей категорией кредитов на образование всего 0.05%. При этом новые данные наглядно показали, что чаще других возвращают в срок кредиты взятые на ремонт.
    
 </div>

## Шаг 4. Общий вывод

В ходе исследования были проанализированы исходные данные заёмщиков по кредитам. В ходе работы было установлено, что данные были частично утеряны при копировании и т.к. общее количество таких данных превычило 10%, они были заменены медианными значения, вычисленными для соответствующих категорий. <br>
Некоторые данные так же имели нетипичные значения в следствие ручного ввода, пропусков и автоматической подстановки значений по умолчанию в пустых полях. В каждом таком случае суммарное количество таких данных составило не более 1% и поэтому влияющие на выводы данные были отфильтрованы. <br>
В работе рассматривались гипотезы влияния наличия детей, семейного статуса, уровня дохода и целей кредита на просрочки по платежам. Несмотря на то, что в некоторых случаях удалось выявить зависимость между описанными факторами и невозвратами кредитов в срок, в целом соотношение возвращённых в срок кредитов к невозвращённым лежит в достаточно узком для всех категорий диапазоне, без выдающихся значений, что может говорить о том, что в имевшейся в распоряжении выборке данных процессы носили статистически равномерный характер без ярко выраженных тенденций.

<div class="alert alert-danger">
<b>❌ Итоговый комментарий ревьюера:</b> 
<br>Ты хорошо потрудился, большая часть работы сделана, но остаётся внести следующие правки:
<br>- изменить тип данных в столбце total_income 
<br>- проверить наличие дубликатов методом duplicated     
<br>- произвести правильные расчеты для ответа на итоговые вопросы
<br>- поправить ответы на итоговые вопросы, если это будет необходимо    
</div>

<div class="alert alert-info">
<b> Комментарий студента:</b> 

- тип данных в столбце total_income изменён на integer <br>
- общая таблица не содержит уникальных данных, таких как: id или ФИО. Остальные данные принимают значения из перечня определённых статусов или имеют случайный характер происхождения, что позволяет им повторяться. Поиску и устранению неявных дубликатов в перечнях статусов посвещён отдельный раздел <br>
- дополнительные расчёты для каждой категории вопросов произведены <br>
- ответы на итоговые вопросы дополнены и скорректированы.
    
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.